ceno_zkvm/src/witness.rs

```diff
@@ -28,98 +29,104 @@ use crate::{
+    pub fn new(num_rows: usize, num_col: usize, padding_strategy: InstancePaddingStrategy) -> Self {
         RowMajorMatrix {
-            values: create_uninit_vec(num_total_rows * num_col),
-            num_padding_rows,
+            values: vec![T::default(); num_rows * num_col],
             num_col,
+            padding_strategy,
         }
     }
 
-    pub fn num_instances(&self) -> usize {
-        self.values.len() / self.num_col - self.num_padding_rows
+    pub fn len(&self) -> usize {
+        self.values.len()
+    }
+
+    pub fn values(&mut self) -> &mut Vec<T> {
+        &mut self.values
     }
 
     pub fn num_padding_instances(&self) -> usize {
-        self.num_padding_rows
+        return next_pow2_instance_padding(self.num_instances()) - self.num_instances();
+    }
+
+    pub fn num_instances(&self) -> usize {
+        self.values.len() / self.num_col
     }
 
-    pub fn iter_rows(&self) -> Chunks<MaybeUninit<T>> {
+    pub fn iter_rows(&self) -> Chunks<T> {
         self.values.chunks(self.num_col)
     }
 
-    pub fn iter_mut(&mut self) -> ChunksMut<MaybeUninit<T>> {
+    pub fn iter_mut(&mut self) -> ChunksMut<T> {
         self.values.chunks_mut(self.num_col)
     }
 
-    pub fn par_iter_mut(&mut self) -> rayon::slice::ChunksMut<MaybeUninit<T>> {
+    pub fn par_iter_mut(&mut self) -> rayon::slice::ChunksMut<T> {
         self.values.par_chunks_mut(self.num_col)
     }
 
-    pub fn par_batch_iter_mut(
-        &mut self,
-        num_rows: usize,
-    ) -> rayon::slice::ChunksMut<MaybeUninit<T>> {
+    pub fn par_batch_iter_mut(&mut self, num_rows: usize) -> rayon::slice::ChunksMut<T> {
         self.values.par_chunks_mut(num_rows * self.num_col)
     }
-
-    pub fn par_batch_iter_padding_mut(
-        &mut self,
-        num_instances: Option<usize>,
-        batch_size: usize,
-    ) -> rayon::slice::ChunksMut<'_, MaybeUninit<T>> {
-        let num_instances = num_instances.unwrap_or(self.num_instances());
-        self.values[num_instances * self.num_col..]
-            .as_mut()
-            .par_chunks_mut(batch_size * self.num_col)
-    }
-
-    pub fn de_interleaving(mut self) -> Vec<Vec<T>> {
-        (0..self.num_col)
-            .map(|i| {
-                self.values
-                    .par_iter_mut()
-                    .skip(i)
-                    .step_by(self.num_col)
-                    .map(|v| unsafe { mem::replace(v, mem::MaybeUninit::uninit()).assume_init() })
-                    .collect::<Vec<T>>()
-            })
-            .collect()
-    }
 }
 
 impl<F: Field> RowMajorMatrix<F> {
     pub fn into_mles<E: ff_ext::ExtensionField<BaseField = F>>(
         self,
     ) -> Vec<DenseMultilinearExtension<E>> {
-        self.de_interleaving().into_mles()
+        let start = Instant::now();
+        let padding_row = match self.padding_strategy {
+            InstancePaddingStrategy::RepeatLast => {
+                self.values[self.values.len() - self.num_col..].to_vec()
```

Comment:
What if empty?

There was a special case `if steps.is_empty()` somewhere:

```rust
                InstancePaddingStrategy::RepeatLast if steps.is_empty() => {
                    tracing::debug!("No {} steps to repeat, using zero padding", Self::name());
                    vec![MaybeUninit::new(E::BaseField::ZERO); num_witin]
                }
```

although that was not completely correct either; better than crashing.